Source/Voidsinger/Voidgrids/Voidgrid.cpp

```diff
@@ -1004,27 +1033,173 @@ void AVoidgrid::AddResources(TMap<EResourceType, float> AddedResources)
+{
+	for (TPair<EResourceType, float> EachResourceTypeToAmountToIncreaseCapacity : ResourceTypesToAmountToIncreaseCapacities)
+	{
+		//Stores the capacity added to this resource type
+		float AddedCapacity = EachResourceTypeToAmountToIncreaseCapacity.Value;
+		//Stores what the new capacity of this resource type will actually be
+		float NewStorageCapacity = ResourceTypesToStorageCapacities.Contains(EachResourceTypeToAmountToIncreaseCapacity.Key) ? ResourceTypesToStorageCapacities.FindRef(EachResourceTypeToAmountToIncreaseCapacity.Key) + AddedCapacity : AddedCapacity;
+		ResourceTypesToStorageCapacities.Emplace(EachResourceTypeToAmountToIncreaseCapacity.Key, NewStorageCapacity);
+	}
+}
+
+/**
+ * Removes storage capacity for each resource type specified, and decreases them by the amount specified
+ *
+ * @param ResourceTypesToAmountToDecreaseCapacities - The resource types to the amount of decreased capacity
+ */
+void AVoidgrid::RemoveResourceStorageCapacity(TMap<EResourceType, float> ResourceTypesToAmountToDecreaseCapacities)
+{
+	for (TPair<EResourceType, float> EachResourceTypeToAmountToDecreaseCapacity : ResourceTypesToAmountToDecreaseCapacities)
+	{
+		if (ResourceTypesToStorageCapacities.Contains(EachResourceTypeToAmountToDecreaseCapacity.Key))
+		{
+			//Stores the current capacity of this resource type
+			float CurrentCapacity = ResourceTypesToStorageCapacities.FindRef(EachResourceTypeToAmountToDecreaseCapacity.Key);
+			//Stores the current capacity minus the amount of capacity that was removed
+			float CurrentCapacityMinusDecreasedAmount = CurrentCapacity - EachResourceTypeToAmountToDecreaseCapacity.Value;
+			//Stores the new storage capacity
+			float NewStorageCapacity = CurrentCapacityMinusDecreasedAmount < 0 ? 0 : CurrentCapacityMinusDecreasedAmount;
+
+			ResourceTypesToStorageCapacities.Emplace(EachResourceTypeToAmountToDecreaseCapacity.Key, NewStorageCapacity);
+		}
+	}
+}
+
+/* /\ Capacity /\ *\
+\* -------------- */
+
+/* ----------- *\
+\* \/ Rates \/ */
+
+/**
+ * Gets the production rates of each resource type
+ *
+ * @return A map of each resource type to how much of each is being produced
+ */
+const TMap<EResourceType, float> AVoidgrid::GetResourceProductionRates() const
+{
+	return ResourceTypesToProductionRates;
+}
+
+/**
+ * Gets the consumption rates of each resource type
+ *
+ * @return A map of each resource type to how much of each is being consumed
+ */
+const TMap<EResourceType, float> AVoidgrid::GetResourceConsumptionRates() const
+{
+	return ResourceTypesToConsumptionRates;
+}
+
+/**
+ * Gets the attempted consumption rates of each resource type. This means it will show what was actually used + what was failed to be used (because there wasn't enough resources).
+ *
+ * @return A map of each resource type to the attempted consumption rate of each
+ */
+const TMap<EResourceType, float> AVoidgrid::GetResourceAttemptedConsumptionRates() const
+{
+	return ResourceTypesToAttemptedConsumptionRates;
+}
+
```

Comment:
what about Get Resource Attempted Production Rates since storage can fail due to full storage

Reply:
The reason I didn't do that is because "production rate" already tracks that. And you can already tell from capacity whether production is being stopped by capacity, because you can just that the capacity is full. But if you want me to add that I can.

Reply:
Yeah, add it anyways.